net/libfab/libfab.c

```diff
@@ -0,0 +1,2963 @@
+
+	M0_ENTRY("loc_buf=%p q=%d loc_seg=%d rem_buf=0x%"PRIx64" rem_seg=%d",
+		 fb, fb->fb_nb->nb_qtype, fb->fb_nb->nb_buffer.ov_vec.v_nr, 
+		 fb->fb_rbd->fbd_bufptr, (int)fb->fb_rbd->fbd_iov_cnt);
+	M0_PRE(fb->fb_rbd != NULL);
+
+	r_iov = fb->fb_riov;
+
+	while(xfer_len < fb->fb_nb->nb_length) {
+		M0_ASSERT(rem_sidx <= fb->fb_rbd->fbd_iov_cnt);
+		loc_slen = v_cnt[loc_sidx] - loc_soff;
+		rem_slen = r_iov[rem_sidx].len - rem_soff;
+		
+		chunk = min64u(loc_slen, rem_slen);
+
+		if (loc_slen > rem_slen) {
+			rem_sidx++;
+			rem_soff = 0;
+			loc_soff += chunk;
+		} else {
+			loc_sidx++;
+			loc_soff = 0;
+			rem_soff += chunk;
+			if(rem_soff >= r_iov[rem_sidx].len) {
+				rem_sidx++;
+				rem_soff = 0;
+			}
+		}
+		wr_cnt++;
+		xfer_len += chunk;
+	}
+	return wr_cnt;
+}
+
+/**
+ * Process bulk operation from transfer machine bulk operation list.
+ */
+static void libfab_bulk_buf_process(struct m0_fab__tm *tm)
+{
+	struct m0_fab__bulk_op *op;
+	uint32_t                available;
+
+	m0_tl_for(fab_bulk, &tm->ftm_bulk, op) {
+		/* Keeping 64 entries on CQ for rpc messages */
+		available = (FAB_MAX_TX_CQ_EV - 64) - op->fbl_aep->aep_bulk_cnt;
+
+		if (op->fbl_buf->fb_wr_cnt <= available &&
+		    op->fbl_aep->aep_tx_state == FAB_CONNECTED) {
+			libfab_bulk_op(op->fbl_aep, op->fbl_buf);
+			fab_bulk_tlist_del(op);
+			op->fbl_buf->fb_bulk_op = NULL;
+			m0_free(op);
+		}
+	} m0_tl_endfor;
+}
+
+/** 
+ * This function will call the bulk transfer operation (read/write) on the
+ * net-buffer.
+ */
+static int libfab_bulk_op(struct m0_fab__active_ep *aep, struct m0_fab__buf *fb)
+{
+	struct fi_msg_rma  op_msg;
+	struct fi_rma_iov *r_iov;
+	m0_bcount_t       *v_cnt = fb->fb_nb->nb_buffer.ov_vec.v_count;
+	m0_bcount_t        xfer_len = 0;
+	struct iovec       iv;
+	uint64_t           op_flag;
+	uint32_t           loc_sidx = 0;
+	uint32_t           rem_sidx = 0;
+	uint32_t           loc_soff = 0;
+	uint32_t           rem_soff = 0;
+	uint32_t           loc_slen;
+	uint32_t           rem_slen;
+	uint32_t           wr_cnt = 0;
+	bool               isread;
+	int                ret = 0;
+	bool               last_seg = false;
+
+	M0_ENTRY("loc_buf=%p q=%d loc_seg=%d rem_buf=0x%"PRIx64" rem_seg=%d",
+		 fb, fb->fb_nb->nb_qtype, fb->fb_nb->nb_buffer.ov_vec.v_nr, 
+		 fb->fb_rbd->fbd_bufptr, (int)fb->fb_rbd->fbd_iov_cnt);
+	M0_PRE(fb->fb_rbd != NULL);
+
+	r_iov = fb->fb_riov;
+	isread = (fb->fb_nb->nb_qtype == M0_NET_QT_ACTIVE_BULK_RECV);
+	fb->fb_wr_comp_cnt = 0;
+
+	while(xfer_len < fb->fb_nb->nb_length) {
+		M0_ASSERT(rem_sidx <= fb->fb_rbd->fbd_iov_cnt);
+		loc_slen = v_cnt[loc_sidx] - loc_soff;
+		rem_slen = r_iov[rem_sidx].len - rem_soff;
+		
+		iv.iov_base = fb->fb_nb->nb_buffer.ov_buf[loc_sidx] + loc_soff;
+		iv.iov_len = min64u(loc_slen, rem_slen);
+
+		if (xfer_len + iv.iov_len >= fb->fb_nb->nb_length)
+			last_seg = true;
+		
+		op_msg.msg_iov       = &iv;
+		op_msg.desc          = &fb->fb_mr.bm_desc[loc_sidx];
+		op_msg.iov_count     = 1;
```

Comment:
As discussed for TCP we can send 4 IOV and for verbs 1. So this can be made like init time config param so if the libfabric supports more in future this code will not change.

if( <TCPEnabled> )
      libfab_max_iov = LIBFAB_MAX_TCP_IOV // The define of libfabric
else // RDMA
      libfab_max_iov = LIBFAB_MAX_RDMA_IOV // The define of libfabric

Reply:
We can take this up during performance tuning for tcp.